internal/cryptokit/hmac.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//go:build darwin
+
+package cryptokit
+
+// #include "cryptokit.h"
+import "C"
+import (
+	"errors"
+	"hash"
+	"runtime"
+	"unsafe"
+)
+
+var _ hash.Hash = (*cryptoKitHMAC)(nil)
+var _ cloneHash = (*cryptoKitHMAC)(nil)
+
+type cryptoKitHMAC struct {
+	pinner runtime.Pinner
+	ptr    unsafe.Pointer
+
+	hashEnum int
+	key      []byte
+
+	blockSize int
+	size      int
+}
+
+func NewHMAC(fh func() hash.Hash, key []byte) hash.Hash {
+	h := fh()
+	if h == nil {
+		return nil
+	}
+
+	hashEnum := hashToHMACEnum(h)
+	if hashEnum == 0 {
+		// The hash function is not supported by the HMAC implementation.
+		return nil
+	}
+
+	pinner := runtime.Pinner{}
+
+	if len(key) > 0 {
+		pinner.Pin(&key[0])
+		defer pinner.Unpin()
+	}
+
+	hmac := &cryptoKitHMAC{
+		pinner: pinner,
+		ptr: C.initMAC(
```

Comment:
You can remove this call here and deduplicate some code by calling `hmac.Reset()` before returning from this function.

Reply:
I can't move finalizer. All I can move is pinning lines in here at best, I think little cloning is better than re-adding this redunant nil check on reset function again.